src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorHelperUpdater.cs

```diff
@@ -0,0 +1,153 @@
+        }
+
+        /// <summary>
+        /// Determines whether this updater needs to run on a given set of RazorCodeDocuments. Will return true if there are @helper functions
+        /// in any of the Razor documents and false otherwise.
+        /// </summary>
+        /// <param name="context">The upgrade context for the currently upgrading solution.</param>
+        /// <param name="inputs">The Razor code documents being upgraded.</param>
+        /// <param name="token">A cancellation token.</param>
+        /// <returns>A FileUpdaterResult with a 'true' value if any documents contain a @helper and false otherwise. If true, the result will also include the paths of the documents containing a @helper.</returns>
+        public async Task<IUpdaterResult> IsApplicableAsync(IUpgradeContext context, ImmutableArray<RazorCodeDocument> inputs, CancellationToken token)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var inputsWithHelpers = new List<string>();
+            foreach (var input in inputs)
+            {
+                // Check whether any of the documents contain a @helper function
+                if (await _helperMatcher.HasHelperAsync(input).ConfigureAwait(false))
+                {
+                    // If a @helper is present, log it to the console and store the path to return later
+                    var path = input.Source.FilePath;
+                    inputsWithHelpers.Add(path);
+                }
+            }
+
+            // Log how many Razor documents had @helpers and return a corresponding FileUpdaterResult
+            _logger.LogInformation("Found @helper functions in {Count} documents", inputsWithHelpers.Count);
+            return new FileUpdaterResult(inputsWithHelpers.Any(), inputsWithHelpers);
+        }
+
+        /// <summary>
+        /// Upgrades a set of RazorCodeDocuments to replace @helper functions with local methods.
+        /// </summary>
+        /// <param name="context">The upgrade context being upgraded.</param>
+        /// <param name="inputs">The Razor code documents to upgrade.</param>
+        /// <param name="token">A cancellation token.</param>
+        /// <returns>A FileUpdaterResult indicating whether the opertaion succeeded and which documents (if any) were updated.</returns>
+        public async Task<IUpdaterResult> ApplyAsync(IUpgradeContext context, ImmutableArray<RazorCodeDocument> inputs, CancellationToken token)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            // Track which documents are updated for inclusion in the updater result returned later
+            var updatedDocuments = new List<string>();
+            foreach (var document in inputs)
+            {
+                var path = document.Source.FilePath;
```

Comment:
I don't know if you can assume they all have a valid path you can read from.

Reply:
I'll check that FilePath is non-null and exists. I don't think there's a better way to get the text, though, because `RazorSourceDocument` doesn't have any public APIs for getting the contents of the document.